src/base/locked_queue_test.cc

```diff
@@ -170,15 +268,71 @@ TEST(LockedQueueIndexTest, ShardIndexGrows) {
   auto inserted_item = list.insert(list.end(), 1);
   index.Put(inserted_item, shard1);
   EXPECT_EQ(shard1 + 1, index.index_.size());
-  EXPECT_EQ(1u, index.index_[shard1].size());
+  EXPECT_EQ(1u, index.index_[shard1].tasks.size());
   EXPECT_EQ(1u, index.reverse_index_.count(&*inserted_item));
 
+  ASSERT_ANY_THROW(index.NotifyShard(shard2));
+
   inserted_item = list.insert(list.end(), 5);
   index.Put(inserted_item, shard2);
   EXPECT_EQ(shard2 + 1, index.index_.size());
-  EXPECT_EQ(1u, index.index_[shard1].size());
-  EXPECT_EQ(1u, index.index_[shard2].size());
+  EXPECT_EQ(1u, index.index_[shard1].tasks.size());
+  EXPECT_EQ(1u, index.index_[shard2].tasks.size());
+  EXPECT_EQ(1u, index.reverse_index_.count(&*inserted_item));
+}
+
+TEST(LockedQueueIndexTest, ShardIndexGrowsOnOverloadedSearch) {
+  List<int> list;
+  LockedQueue<int>::Index index;
+
+  const ui32 shard1 = 4u;
+  const ui32 shard2 = 5u;
+
+  auto inserted_item = list.insert(list.end(), 1);
+  index.Put(inserted_item, shard1);
+  EXPECT_EQ(shard1 + 1, index.index_.size());
+  EXPECT_EQ(1u, index.index_[shard1].tasks.size());
   EXPECT_EQ(1u, index.reverse_index_.count(&*inserted_item));
+
+  ASSERT_ANY_THROW(index.ShardIsEmpty(shard2));
+  ASSERT_ANY_THROW(index.GetStrict(shard2));
+
+  const ui32 max_queue_size = 3u;
+
+  EXPECT_EQ(shard2, index.MaybeOverloadedShard(max_queue_size, shard2));
+  EXPECT_EQ(shard2 + 1, index.index_.size());
+  EXPECT_EQ(1u, index.index_[shard1].tasks.size());
+  EXPECT_EQ(1u, index.reverse_index_.count(&*inserted_item));
+  EXPECT_EQ(0u, index.index_[shard2].tasks.size());
+}
+
+TEST(LockedQueueIndexTest, MaybeOverloadedReturnsOverloadedShard) {
```

Comment:
Don't forget to check that `MaybeOverloaded` considers initial shard non-emptiness